### PR TITLE
Checkpoint Ablation for Flamby

### PR DESCRIPTION
**Pull request by emersodb**
Adding in the ability to perform "non-federated checkpointing" for the FLamby experiments. This facilitates the checkpoint ablation study that was added to the ICLR paper.

Fixed two small bugs as well. The first was the dataloader method name for APFL (had a typo) and the second was defaulting the auxiliary FENDA losses to zeros rather than `None`. This prevents using non-symmetric latent spaces in FENDA, because you need symmetric latent spaces for some the auxiliary losses and they are computed even if the coefficients are zero.

### CODE REVIEW

File: research/flamby/fed_ixi/apfl/client.py [hunk 1]
```diff
@@ -50,7 +54,7 @@ def __init__(
         self.alpha_learning_rate = alpha_learning_rate
         self.client_number = client_number
 
-    def get_dataloader(self, config: Config) -> Tuple[DataLoader, DataLoader]:
```
Comment (emersodb) [inline research/flamby/fed_ixi/apfl/client.py:53]: Somehow a typo crept into this APFL client

File: fl4health/clients/fenda_client.py [hunk 2]
```diff
@@ -23,9 +23,9 @@ def __init__(
         metric_meter_type: MetricMeterType = MetricMeterType.AVERAGE,
         checkpointer: Optional[TorchCheckpointer] = None,
         temperature: Optional[float] = 0.5,
-        perfcl_loss_weights: Optional[Tuple[float, float]] = (0.0, 0.0),
-        cos_sim_loss_weight: Optional[float] = 0.0,
-        contrastive_loss_weight: Optional[float] = 0.0,
```
Comment (emersodb) [inline fl4health/clients/fenda_client.py:28]: If these are not None, then some of these losses (at least perfcl) are still computed, but just "zeroed" out. This is problematic if you want to use asymmetric latent spaces in a FENDA architecture.

File: research/flamby/fed_heart_disease/apfl/client.py [hunk 3]
```diff
@@ -109,17 +113,28 @@ def get_criterion(self, config: Config) -> _Loss:
     parser.add_argument(
         "--alpha_learning_rate", action="store", type=float, help="Learning rate for the APFL alpha", default=0.01
     )
+    parser.add_argument(
+        "--no_federated_checkpointing",
+        action="store_true",
+        help="boolean to indicate whether we're evaluating an APFL model or not, as those model have special args",
```
Comment (yc7z) [inline research/flamby/fed_heart_disease/apfl/client.py:119]: Is this the right helper message?

File: research/flamby/fed_heart_disease/fenda/client.py [hunk 4]
```diff
@@ -116,16 +120,27 @@ def get_criterion(self, config: Config) -> _Loss:
     parser.add_argument("--cos_sim_loss", action="store_true", help="Activate Cosine Similarity loss")
     parser.add_argument("--contrastive_loss", action="store_true", help="Activate Contrastive loss")
     parser.add_argument("--perfcl_loss", action="store_true", help="Activate PerFCL loss")
+    parser.add_argument(
+        "--no_federated_checkpointing",
+        action="store_true",
+        help="boolean to indicate whether we're evaluating an APFL model or not, as those model have special args",
```
Comment (yc7z) [inline research/flamby/fed_heart_disease/fenda/client.py:126]: Also the helper message here

File: research/flamby/fed_ixi/apfl/client.py [hunk 5]
```diff
@@ -111,17 +115,28 @@ def get_criterion(self, config: Config) -> _Loss:
     parser.add_argument(
         "--alpha_learning_rate", action="store", type=float, help="Learning rate for the APFL alpha", default=0.01
     )
+    parser.add_argument(
+        "--no_federated_checkpointing",
+        action="store_true",
+        help="boolean to indicate whether we're evaluating an APFL model or not, as those model have special args",
```
Comment (yc7z) [inline research/flamby/fed_ixi/apfl/client.py:121]: Also the helper message here

File: research/flamby/fed_ixi/fenda/client.py [hunk 6]
```diff
@@ -117,16 +120,27 @@ def get_criterion(self, config: Config) -> _Loss:
     parser.add_argument("--cos_sim_loss", action="store_true", help="Activate Cosine Similarity loss")
     parser.add_argument("--contrastive_loss", action="store_true", help="Activate Contrastive loss")
     parser.add_argument("--perfcl_loss", action="store_true", help="Activate PerFCL loss")
+    parser.add_argument(
+        "--no_federated_checkpointing",
+        action="store_true",
+        help="boolean to indicate whether we're evaluating an APFL model or not, as those model have special args",
```
Comment (yc7z) [inline research/flamby/fed_ixi/fenda/client.py:126]: Helper message in parser

**Comment by yc7z** [comment]
Looks good to me. Only a few comments on some small typos

**Review by yc7z — APPROVED** [review]

**Comment by emersodb** [comment]
> Looks good to me. Only a few comments on some small typos

Thanks for pointing out the helper message. Will fix!

**Pull request merged**